src/include/kernel_utils/hugetlb.h

```diff
@@ -0,0 +1,139 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* X-SPDX-Copyright-Text: (c) Copyright 2023 Xilinx, Inc. */
+/*
+ ****************************************************************************
+ *
+ * A library to allocate hugepages in the kernelspace
+ * (which is not encouraged and not straightforward).
+ *
+ * How it works: the userspace donates a memfd file to the kernel module via
+ * ioctl(). Then the library calls ftruncate(), fallocate(), mmap(),
+ * pin_user_pages() and munmap() to get an instance of a 2 MiB hugepage,
+ * which we can give to NIC after dma_map_single(). For older kernels without
+ * the memfd support, the library calls hugetlb_file_setup() directly.
+ *
+ * Ideally, the userspace should do the allocation itself, i.e. ftruncate(),
+ * fallocate(), mmap(), and then make an ioctl() with the virtual address.
+ * In this case, we become aligned with the kernel API, and the library
+ * degenerates into a single pin_user_pages() call. This may require
+ * significant workflow refactoring in Onload.
+ *
+ ****************************************************************************
+ */
+
+#ifndef __HUGETLB_H__
+#define __HUGETLB_H__
+
+#include <onload/common.h>
+#include <onload/atomics.h>
+
+#define HUGEPAGE_SIZE (2 * 1024 * 1024)
+
+struct hugetlb_allocator {
+	struct file *filp;
+	off_t offset;
+	atomic_t refcnt;
+};
+
+struct hugetlb_page {
```

Comment:
I dislike the names.  10 years later, Linux kernel will introduce `hugetlb_page` structure, and we'll get a new compat issue out of nowhere.  Please add `oo_` prefix to all your names, or something like this.

Reply:
I'm on the same page here. I didn't like the "crude" `hugetlb_xyz` names but couldn't think of a good prefix. Please find these prefixed with `oo_` in the latest PR incarnation.